Feature/jetbrains ide companion

SoLoHiC opened the pull request:
## TLDR
Really liked Gemini-CLI and can't wait for it working within IntelliJ IDEA and other JetBrains IDE. So I took a few days during this vacation working on this plugin(, well, most of the code is actually written by Gemini itself and I'm just co-piloting along , dealing with a few tricky issues about how the MCP's streamable http transport works in Kotlin and making sure its UI works as expected.

This PR creates a JetBrains plugin providing similar features like the VS Code companion extension does.  As I've tested locally, it should basically follow the https://github.com/google-gemini/gemini-cli/blob/main/docs/ide-companion-spec.md.  

## Dive Deeper
All code changes can be found under packages/jebrains-ide-companion/ .

## Reviewer Test Plan
the way i tested it after pulling the code:

1. go to the project root directory
2. Build the plugin and the customized gemini-cli using `npm run build`
3. In your JetBrains IDE, go to Settings > Plugins > Install Plugin from Disk
4. Select the generated plugin ZIP file from `packages/jebrains-ide-companion/build/distributions/`
5. go back to the project root directory, install the customized gemini-cli using `npm install -g .`
6. run `gemini` in the JetBrains IDE's Terminal window
7. run `/ide status` within gemini-cli, should be showing ' Connected to IntelliJ IDEA 2025.2'

And also, other main features like the VSCode extension provided can be test too. Demo result below:
<img width="2880" height="1794" alt="image" src="https://github.com/user-attachments/assets/64057bed-d258-45ff-a92e-8c4d8afcc47d" />


## Testing Matrix

<!-- Before submitting please validate your changes on as many of these options as possible -->

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✅  | ❓  | ❓  |
| npx      | ❓  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ❓  | -   | -   |

## Linked issues / bugs

I see there already is an issue created by [alcarraz](https://github.com/alcarraz), (https://github.com/google-gemini/gemini-cli/issues/9273)
[skeshive](https://github.com/skeshive) , please take a look at this and thanks for your work on the VS Code extension.

<!--
Link to any related issues or bugs.

**If this PR fully resolves the issue, use one of the following keywords to automatically close the issue when this PR is merged:**

- Closes #<issue_number>
- Fixes #<issue_number>
- Resolves #<issue_number>

*Example: `Resolves #123`*

**If this PR is only related to an issue or is a partial fix, simply reference the issue number without a keyword:**

*Example: `This PR makes progress on #456` or `Related to #789`*
-->

kairosci commented:
If you are able to do, in commit, could you also add a button that suggests commit messages, like GitHub Copilot does? So you only have one thing.

reconsumeralization commented:
Does it maintain secure validated https and sse between extansion/ide extension/agent agent/anything with verified by each agent on all communication of source?

diver80 commented:
Unsure if this is the correct place (otherwise feel free to delete comment) but here are my testing results. Any help is appreciated.

**setup**
- OS Linux - Ubuntu 22.04
- IntelliJ IDEA 2025.2.2 (Ultimate Edition))
- gemini settings disableAutoUpdate: true

**Building**
- successfull
- after small modification (see https://github.com/SoLoHiC/gemini-cli/pull/1) plugin zip file was created successfully

**Installation**
- zip was successfully installed in IntelliJ and icons for starting the cli are shown

**Using**
- Clicking on the icons in IntelliJ gemini is correctly started but connection with the integration isn't possible
- Logs during startup:
`
(node:1024149) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
(node:1024169) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
Loaded cached credentials.
[ERROR] [IDEClient] Failed to connect to IDE companion extension. Please ensure the extension is running. To install the extension, run /ide install.
`

**What was tested**
- /ide status
`
✕ 🔴 Disconnected: Failed to connect to IDE companion extension. Please ensure the extension is running. To install the extension, run /ide install.
`
- /ide install => no effect as zip was already installed

skeshive commented on packages/core/src/ide/detect-ide.ts:
you should just be able to set these values in the port file and not have to edit this file at all

lmk if that doesn't work

SoLoHiC commented:
> @SoLoHiC sorry for the delay- could you share more info on how IJ sets the process id for the plugin? I can update the process detection logic in a separate PR so we don't need to make any changes in this one... the file should be the primary source of truth for the discovery

yah @skeshive , it's fair enough to isolate the changes among process detection from this one. I'll revert the 2 commits change of the Gemini CLI core side.

About how IJ set process id...I'm sorry that i don't really have a solid understanding upon it, especially when i've never really run a IJ IDE bundled with gemini plugin on Linux OS. All i could offer is that i reproduced similar detection issue when i ran plugin in debug mode (, starting a intellij community IDE via Intellij IDEA's gradle run command) and then ran gemini cli within the Intellij community IDE. I looked through log that gemini cli found the pid linked to the gradle daemon, while plugin of a intellij community IDE wrote process info using pid of IDE itself. 

pic 1st, log info of pid found of both dev mode ide printed and gemini cli detected and printed
<img width="2358" height="1572" alt="image" src="https://github.com/user-attachments/assets/16eef8d3-e4e8-49ae-8321-5169c63a8947" />

pic 2nd, pid that gemini cli detected links to gradle daemon process, which is how i started the whole "intellij community IDE"
<img width="2876" height="1444" alt="image" src="https://github.com/user-attachments/assets/760b74c8-a224-45d4-865e-0472326a3758" />

pic 3rd, pid that intellij community IDE used to name discovery file links to ides itself.
<img width="2876" height="1652" alt="image" src="https://github.com/user-attachments/assets/f0120794-7350-47bc-8aae-0ca724f34f20" />

hope these debug info helps.

skeshive commented:
Thanks @SoLoHiC, I filed a separate bug for the IDE process detection bug (https://github.com/google-gemini/gemini-cli/issues/12089)

SoLoHiC commented:
> I've manually tested and confirmed that the IDE process ID detection is working correctly for IntelliJ. (Ref: [Issue Comment](https://github.com/google-gemini/gemini-cli/issues/12089#issuecomment-3453110665))
> 
> I also isolated the companion plugin connection issue:
> 
> * Fails: When running in the sandbox via ./gradlew runIde.
> * Succeeds: When the plugin is bundled (./gradlew buildPlugin) and installed from disk.
> 
> Given that users will be installing the plugin from the marketplace, the sandbox environment failure is not a blocker.
> 
> Overall the plugin works great. This is a huge win for IJ users! 🥳
> 
> One small piece of feedback would be to see if the focus can remain in the CLI when a diff is opened by the plugin. That way users can press enter to accept the diff without having to click back into the CLI UI

@skeshive i've fixed the focus issue: when gemini cli opening a diff view window, keep window focus at where user were, which means when user were at CLI terminal window, focus would remain and user could directly press enter to acccept diff😉. 

also, rebased the main branch.

SoLoHiC commented:
updated package-lock.json file to sync jetbrains-ide-companion package info, solving 'npm ci' checks failure.

ignatov commented:
@SoLoHiC Thanks a lot for you contribution!

ignatov commented:
@SoLoHiC As I said before, thanks for your work! Keep in mind that at JetBrains we’re planning our work for the next sprint, and supporting agents in our terminal as well as integrating agents better is on our plate.
